File: backend/src/main/java/com/unina/ratatouille23/menu/entity/Elemento.java
package com.unina.ratatouille23.menu.entity;

import com.fasterxml.jackson.annotation.JsonBackReference;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;

@Entity
public class Elemento {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    @Column(name = "id_elemento", unique = true)
    private Integer id;

    @Column(nullable = false)
    private String nome;

    @Column(nullable = false)
    private String descrizione;

    @Column(nullable = false)
    private double costo;

    @Column(nullable = false)
    private String allergeni;
    
    @ManyToOne
    @JoinColumn(name = "fk_id_categoria")
    @JsonBackReference
    private Categoria categoria;

    public Elemento(Integer id, String nome, String descrizione, double costo, String allergeni, Categoria categoria) {
        this.id = id;
        this.nome = nome;
        this.descrizione = descrizione;
        this.costo = costo;
        this.allergeni = allergeni;
        this.categoria = categoria;
    }
    

    public Elemento() {
    }

    public Categoria getCategoria() {
        return this.categoria;
    }

    public void setCategoria(Categoria categoria) {
        this.categoria = categoria;
    }

    public int getId() {
        return this.id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getNome() {
        return this.nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public String getDescrizione() {
        return this.descrizione;
    }

    public void setDescrizione(String descrizione) {
        this.descrizione = descrizione;
    }

    public double getCosto() {
        return this.costo;
    }

    public void setCosto(double costo) {
        this.costo = costo;
    }

    public String getAllergeni() {
        return this.allergeni;
    }

    public void setAllergeni(String allergeni) {
        this.allergeni = allergeni;
    }


}

File: client/lib/models/Elemento.dart
class Elemento{

  late int _id;
  late String _nome;
  late String _descrizione;
  late double _costo;
  late String _allergeni;
  late int _id_categoria;

  Elemento(int id, String nome, String descrizione, double costo, String allergeni, int id_categoria){
    _id=id;
    _nome=nome;
    _descrizione=descrizione;
    _costo=costo;
    _allergeni=allergeni;
    _id_categoria=id_categoria;
  }

  Elemento.fromJson(Map<String, dynamic> json):
    _id=json['id'],
    _nome=json['nome'],
    _descrizione=json['descrizione'],
    _costo=json['costo'],
    _allergeni=json['allergeni'],
    _id_categoria=json['id_categoria'];

  Map<String, dynamic> toJson(){
    return{
      'id':_id,
      'nome':_nome,
      'descrizione':_descrizione,
      'costo':_costo,
      'allergeni':_allergeni,
      'id_categoria':_id_categoria
    };
  }


  void set_id(int id){_id=id;}
  void set_nome(String nome){_nome=nome;}
  void set_descrizione(String descrizione){_descrizione=descrizione;}
  void set_costo(double costo){_costo=costo;}
  void set_allergeni(String allergeni){_allergeni=allergeni;}
  void set_id_categoria(int id_categoria){_id_categoria=id_categoria;}

  int get_id(){return _id;}
  String get_nome(){return _nome;}
  String get_descrizone(){return _descrizione;}
  double get_costo(){return _costo;}
  String get_allergeni(){return _allergeni;}
  int get_id_categoria(){return _id_categoria;}


}